source/containers.tex

```diff
@@ -20957,3 +21022,628 @@
+\indexlibraryglobal{\exposid{first_}}%
+\begin{itemdecl}
+template<class T>
+  constexpr T @\exposid{de-ice}@(T val) { return val; }
+template<@\exposconcept{integral-constant-like}@ T>
+  constexpr auto @\exposid{de-ice}@(T) { return T::value; }
+
+template<class IndexType, size_t k, class ... SliceSpecifiers>
+  constexpr IndexType @\exposid{first_}@(SliceSpecifiers... slices);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\mandates
+\tcode{IndexType} is a signed or unsigned integer type.
+
+%FIXME: What is $k$ in this itemdescr? Did we mean the \tcode{k} passed to first_? If so, what do we do for de-ice?
+\pnum
+Let $\phi{}_k$ denote the following value:
+\begin{itemize}
+\item
+$s_k$ if $S_k$ models \tcode{\libconcept{convertible_to}<IndexType>}; otherwise
+\item
+\tcode{get<0>($s_k$)}
+if $S_k$ models \tcode{\exposconcept{index-pair-like}<IndexType>}; otherwise
+\item
+\tcode{\exposid{de-ice}($s_k$.offset)}
+if $S_k$ is a specialization of \tcode{strided_slice}; otherwise
+\item
+\tcode{0}.
+\end{itemize}
+
+\pnum
+\expects
+$\phi{}_k$ is representable as a value of type \tcode{IndexType}.
+
+\pnum
+\returns
+\tcode{extents<IndexType>::\exposid{index-cast}($\phi{}_k$)}.
+\end{itemdescr}
+
+\indexlibraryglobal{\exposid{last_}}%
+\begin{itemdecl}
+template<size_t k, class Extents, class ... SliceSpecifiers>
+  constexpr auto @\exposid{last_}@(const Extents& src, SliceSpecifiers... slices);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\mandates
+\tcode{Extents} is a specialization of \tcode{extents}.
+
+\pnum
+Let \tcode{index_type} name the type \tcode{typename Extents::index_type}.
+
+%FIXME: What is $k$ in this itemdescr? Don't we want \tcode{k} and not $k$?
+\pnum
+Let $\lambda{}_k$ denote the following value:
+\begin{itemize}
+\item
+\tcode{\exposid{de-ice}($s_k$) + 1}
+if $S_k$ models \tcode{\libconcept{convertible_to}<index_type>}; otherwise
+\item
+\tcode{get<1>($s_k$)}
+if $S_k$ models \tcode{\exposconcept{index-pair-like}<index_type>}; otherwise
+\item
+\tcode{\exposid{de-ice}($s_k$.offset)} \tcode{+}
+\tcode{\exposid{de-ice}($s_k$.extent)}
+if $S_k$ is a specialization of \tcode{strided_slice}; otherwise
+\item
+%FIXME: Note that the paper uses \tcode{k} here.
+\tcode{src.extent(k)}.
+\end{itemize}
+
+\pnum
+\expects
+$\lambda{}_k$ is representable as a value of type \tcode{index_type}.
+
+\pnum
+\returns
+\tcode{Extents::\exposid{index-cast}($\lambda{}_k$)}.
+\end{itemdescr}
+
+\indexlibraryglobal{\exposid{src-indices}}%
+\begin{itemdecl}
+template<class IndexType, size_t N, class ... SliceSpecifiers>
+  constexpr array<IndexType, sizeof...(SliceSpecifiers)>
+    @\exposid{src-indices}@(const array<IndexType, N>& indices, SliceSpecifiers ... slices);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\mandates
+\tcode{IndexType} is a signed or unsigned integer type.
+
+\pnum
+\returns
+An \tcode{array<IndexType, sizeof...(SliceSpecifiers)> src_idx} such that
+for each $k$ in the range \range{0}{sizeof...(SliceSpecifiers)},
```

Comment:
It was not an accident. The /k/ here is just for the ranging index and has no scope beyond the scope of this returns clause. We tend to use math fonts for indexes like this so I did.

What we have above was left as is because I was confused about which k we were talking about - was it supposed to be a localized index or the parameter?